[Fleet] update get package info API to support package level permissions

joeypoon opened the pull request:
## Summary

Update get package info API to support package level permissions. If user has no fleet permissions but has package level permissions (currently only supporting endpoint), the API will now return the package. Also includes some minor logic consolidation for package permission checking in package policy handler.


### Checklist

- [x] [Unit or functional tests](https://www.elastic.co/guide/en/kibana/master/development-tests.html) were updated or added to match the most common scenarios


### For maintainers

- [x] This was checked for breaking API changes and was [labeled appropriately](https://www.elastic.co/guide/en/kibana/master/contributing.html#kibana-release-notes-process)

elasticmachine commented:
Pinging @elastic/fleet (Team:Fleet)

nchaulet reviewed (approved):
Looks good to me 🚀  and worked as expected

paul-tavares commented on x-pack/plugins/fleet/server/routes/epm/handlers.ts:
```diff
@@ -87,12 +87,16 @@ export const getListHandler: FleetRequestHandler<
       ...request.query,
     });
+
+    const items = filterAllowedPackages(res, 'id', limitedToPackages);
```
so up until this change, we have been failing the API if the data contained more than what they are allowed to see. This change is now filtering out packages instead of failing the request, so that behaviour differs from that approach. I would like to see if we can stay with that model before returning data filtered out. 

⚠️ The bigger issue I see is that this response will be "weird" because of the pagination query params - ex. user requested 10 per page and lets say there are 100 total items. They would potentially get back a response that indicates the total is like `100` but the content of this one page could have less than the `perPage` value in the query params. That's not a good solution. 

I just checked the EPM APIs from fleet and it looks like there is now an API that we can use to retrieve only the Endpoint package - `/api/fleet/epm/packages/endpoint`. I think [we should change the implementation on the security solution](https://github.com/elastic/kibana/blob/main/x-pack/plugins/security_solution/public/management/services/policies/ingest.ts#L136-L148) side to use this API instead and update that API (not this one) to support package privileges . 


**Some History:**

If I remember correctly, we (in security solution) only need to get the version number of the Endpoint package in order to be able to create a URL jump to the integration Package detail page (ex. `http://localhost:5601/app/integrations/detail/endpoint-8.6.0/overview`). The only available way to do that at the time (FYI: FLeet was not even released back then, so everything was very new) was via the List EPM api using a filter of `category=security` and then finding the Endpoint package in that list. We do this because:

1. The API here does not have a way to filter by `package.name`
2. The integrations details page does not support using only the `package.name` in the URL

But recognizing that we now seem to have a better api, lets update our code to use that instead.

joeypoon replied:
In light of this context, switching to `/api/fleet/epm/packages/endpoint` definitely sounds better.

A bit moot since I agree we should just switch APIs but the reason I had to take a filter approach instead of 403 immediately on unauthorized package is due to the fact that this API would also 403 since, as you mentioned, we can't filter by package name.

paul-tavares reviewed (changes requested):
Left some comments around the approach taken on the list of EPM packages and the fact that we probably should no longer be using that API from the security solution side and instead should be calling a different one.

joeypoon commented on x-pack/plugins/fleet/server/services/check_allowed_packages.ts:
```diff
@@ -0,0 +1,71 @@
+
+  const allowedPackagedSet = new Set(allowedPackages);
+  const allowedPackagesStr = allowedPackages.join(', ');
+  const restrictedPackages = packages.reduce((acc, pkg) => {
+    const pkgName = get(pkg, keyPath, undefined);
+    if (!pkgName) {
+      throw new FleetUnauthorizedError(
+        `Authorization denied. Allowed package(s): ${allowedPackagesStr}.`
+      );
+    }
+
+    const isRestricted = !allowedPackagedSet.has(pkgName);
+    if (!isRestricted) return acc;
+
+    return [...acc, pkgName];
+  }, [] as string[]);
```
I was thinking this would make for a better UX so users can request all permissions from admin in one swoop but I can see how it could be a security concern and it being potentially large is certainly another issue. Easy enough to change 👌.

joeypoon commented on x-pack/plugins/fleet/server/services/security/route_required_authz.ts:
```diff
@@ -126,6 +128,39 @@ const ROUTE_AUTHZ_REQUIREMENTS = deepFreeze<Record<string, FleetRouteRequiredAut
+      integrations: {
+        readPackageInfo: true,
+      },
+      packagePrivileges: {
+        endpoint: {
+          actions: {
+            readEndpointList: {
+              executePackageAction: true,
+            },
+            readPolicyManagement: {
+              executePackageAction: true,
+            },
+            readTrustedApplications: {
+              executePackageAction: true,
+            },
+            readEventFilters: {
+              executePackageAction: true,
+            },
+            readHostIsolationExceptions: {
+              executePackageAction: true,
+            },
+            readBlocklist: {
+              executePackageAction: true,
+            },
+          },
+        },
```
We're making the call on security solution visit so the idea here is to make sure if they can access any of the pages, they can make this call. But, thinking about it more, this should probably just be `readSecuritySolution`. It's not currently exposed on fleetAuthz but is simple enough to add.

joeypoon commented:
Updated logic on security solution side to call `GET /api/fleet/epm/packages/endpoint` instead of `GET /api/fleet/epm/packages?category=security`. Also updated `GET /api/fleet/epm/packages/endpoint` API to support package level permissions.

joeypoon commented on x-pack/plugins/security_solution/common/endpoint/service/authz/authz.ts:
```diff
@@ -280,66 +286,3 @@ export const getEndpointAuthzInitialState = (): EndpointAuthz => {
   };
 };
-
-const SIEM_PERMISSIONS = [
```
`canWriteSecuritySolution` and `canReadSecuritySolution` will now come from `fleetAuthz` since we needed access to these within fleet.

joeypoon commented on x-pack/plugins/fleet/server/routes/epm/index.ts:
```diff
@@ -186,9 +192,11 @@ export const registerRoutes = (router: FleetAuthzRouter) => {
-      fleetAuthz: {
-        integrations: { readPackageInfo: true },
-      },
+      fleetAuthz: (fleetAuthz: FleetAuthz): boolean =>
```
didn't want to update this since it's marked deprecated but route matching will always hit this when optional `packageVersion` is omitted from path.

elasticmachine commented:
Pinging @elastic/security-onboarding-and-lifecycle-mgt (Team:Onboarding and Lifecycle Mgt)

dasansol92 commented on x-pack/plugins/security_solution/common/endpoint/service/authz/authz.ts:
```diff
@@ -70,13 +62,26 @@ export const calculateEndpointAuthz = (
+    fleetAuthz,
+    isEndpointRbacEnabled,
+    hasEndpointManagementAccess,
+    'writeSecuritySolution'
```
Does this and the one below needs the `isEndpointRbacEnabled` check? This privilege already existed before, right?

joeypoon replied:
you're right, will update.

joeypoon commented on x-pack/plugins/security_solution/common/endpoint/generate_data.ts:
```diff
@@ -1657,6 +1659,90 @@ export class EndpointDocGenerator extends BaseDataGenerator {
+  /**
+   * Generate a Fleet EPM Package for Endpoint
+   */
+  public generateEpmPackageInfo(): GetInfoResponse['item'] {
```
I agree. I initially tried updating `generateEpmPackage` but it adds a bunch of additional file touches since the returned type is different. I think it's best to have a separate PR for removing `/api/fleet/epm/packages?category=security` related code from endpoint since this PR has already gotten quite large.

ashokaditya reviewed (approved):
Checked it out again and it works as expected. 🚢  it.

paul-tavares commented on x-pack/plugins/fleet/server/services/check_allowed_packages.ts:
```diff
@@ -0,0 +1,48 @@
+export function checkAllowedPackages<T>(
+  packages: T[],
+  allowedPackages?: string[],
+  keyPath: string = ''
```
Nice 👍

paul-tavares commented on x-pack/plugins/security_solution/public/management/services/policies/ingest.ts:
```diff
@@ -135,12 +137,10 @@ export const sendGetFleetAgentStatusForPolicy = (
-  );
+): Promise<GetInfoResponse['item']> => {
+  const path = epmRouteService.getInfoPath('endpoint');
+  const securityPackages = await http.get<GetInfoResponse>(path);
```
Maybe rename the variable to `endpointPackageResponse`?

paul-tavares commented on x-pack/plugins/security_solution/scripts/endpoint/agent_emulator/services/endpoint_loader.ts:
```diff
@@ -105,7 +105,7 @@ export const loadEndpoints = async ({
   }
 
-  const endpointPackage = await fetchEndpointPackageInfo(kbnClient);
+  const endpointPackage = await getEndpointPackageInfo(kbnClient);
```
nice. Thank you.

joeypoon commented on x-pack/plugins/fleet/common/constants/authz.ts:
```diff
@@ -5,22 +5,123 @@
+
+const SECURITY_SOLUTION_APP_ID = 'siem';
+
+export const ENDPOINT_PRIVILEGES = {
```
added some comments but basically needed to add some mappings since `readSecuritySolution` and `writeSecuritySolution` aren't fetched from kibana quite the same as the other privileges: https://github.com/joeypoon/kibana/blob/4c8dd903f6ae89941bb60dda5428af6da9a27b7e/x-pack/plugins/fleet/server/services/security/security.ts#L71

joeypoon commented on x-pack/test/fleet_api_integration/apis/test_users.ts:
```diff
@@ -128,6 +128,16 @@ export const testUsers: {
     username: 'endpoint_fleet_read_integr_none',
     password: 'changeme',
   },
+  endpoint_integr_read_only_no_fleet: {
```
renamed to `endpoint_integr_read_only_fleet_none` for consistency with the others. it does have `siem: ['minimal_all']` so kept the endpoint read only in name.
